File: leetcode-daily/march-2024/205-isomorphic-strings.cpp
/*
https://leetcode.com/problems/isomorphic-strings

Given two strings `s` and `t`, *determine if they are isomorphic*.

Two strings `s` and `t` are isomorphic if the characters in `s` can be replaced to get `t`.

All occurrences of a character must be replaced with another character while preserving the order of
characters. No two characters may map to the same character, but a character may map to itself.


**Example 1:**

**Input:** s = "egg", t = "add"
**Output:** true

**Example 2:**

**Input:** s = "foo", t = "bar"
**Output:** false

**Example 3:**

**Input:** s = "paper", t = "title"
**Output:** true


**Constraints:**

* `1 <= s.length <= 5 * 10⁴`
* `t.length == s.length`
* `s` and `t` consist of any valid ascii character.
*/

#include "./../../cpp-utils/printutils.h"
#include "bits/stdc++.h"
#include <iostream>
using namespace std;

class Solution {
  public:
    bool isIsomorphic(string s, string t) {
        int n = s.length();
        if (n != t.length())
            return false;

        vector<char> s_t_mapping(128, '\0');
        vector<bool> t_mapped(128, false);
        vector<bool> s_mapped(128, false);
        string s_new = s;
        for (int i = 0; i < n; ++i) {
            if (s_t_mapping[s[i]] == '\0') {
                if (t_mapped[t[i]])
                    return false;
                if (s_mapped[s[i]])
                    return false;
                s_t_mapping[s[i]] = t[i];
                t_mapped[t[i]] = true;
                s_mapped[s[i]] = true;
            }
            s_new[i] = s_t_mapping[s[i]];
        }
        return s_new == t;
    }
};

int main() {
    Solution sol;

    string s_1 = "egg";
    string t_1 = "add";
    bool expected_1 = true;
    bool output_1 = sol.isIsomorphic(s_1, t_1);
    cout << output_1 << endl;
    assert(output_1 == expected_1);

    string s_2 = "foo";
    string t_2 = "bar";
    bool expected_2 = false;
    bool output_2 = sol.isIsomorphic(s_2, t_2);
    cout << output_2 << endl;
    assert(output_2 == expected_2);

    string s_3 = "paper";
    string t_3 = "title";
    bool expected_3 = true;
    bool output_3 = sol.isIsomorphic(s_3, t_3);
    cout << output_3 << endl;
    assert(output_3 == expected_3);

    return 0;
}

File: leetcode-daily/april-2024/205-isomorphic-strings.rs
/*
https://leetcode.com/problems/isomorphic-strings

Given two strings `s` and `t`, *determine if they are isomorphic*.

Two strings `s` and `t` are isomorphic if the characters in `s` can be replaced to get `t`.

All occurrences of a character must be replaced with another character while preserving the order of characters. No two
characters may map to the same character, but a character may map to itself.


**Example 1:**

**Input:** s = "egg", t = "add"
**Output:** true

**Example 2:**

**Input:** s = "foo", t = "bar"
**Output:** false

**Example 3:**

**Input:** s = "paper", t = "title"
**Output:** true


**Constraints:**

* `1 <= s.length <= 5 * 10⁴`
* `t.length == s.length`
* `s` and `t` consist of any valid ascii character.
*/

struct Solution;
impl Solution {
    pub fn is_isomorphic(s: String, t: String) -> bool {
        let n: usize = s.len();
        if t.len() != n {
            return false;
        }

        let s_chars: Vec<char> = s.chars().collect();
        let t_chars: Vec<char> = t.chars().collect();
        let mut s_t_mapping: Vec<char> = vec!['\0'; 128];
        let mut s_mapped: Vec<bool> = vec![false; 128];
        let mut t_mapped: Vec<bool> = vec![false; 128];
        for i in 0..n {
            if s_t_mapping[s_chars[i] as usize] == '\0' {
                if s_mapped[s_chars[i] as usize] {
                    return false;
                }
                if t_mapped[t_chars[i] as usize] {
                    return false;
                }
                s_t_mapping[s_chars[i] as usize] = t_chars[i];
                s_mapped[s_chars[i] as usize] = true;
                t_mapped[t_chars[i] as usize] = true;
            }

            if s_t_mapping[s_chars[i] as usize] != t_chars[i] {
                return false;
            }
        }

        true
    }
}

fn main() {
    let s_1: String = "egg".to_string();
    let t_1: String = "add".to_string();
    let expected_1: bool = true;
    let output_1: bool = Solution::is_isomorphic(s_1, t_1);
    println!("{:?}", output_1);
    assert!(output_1 == expected_1);

    let s_2: String = "foo".to_string();
    let t_2: String = "bar".to_string();
    let expected_2: bool = false;
    let output_2: bool = Solution::is_isomorphic(s_2, t_2);
    println!("{:?}", output_2);
    assert!(output_2 == expected_2);

    let s_3: String = "paper".to_string();
    let t_3: String = "title".to_string();
    let expected_3: bool = true;
    let output_3: bool = Solution::is_isomorphic(s_3, t_3);
    println!("{:?}", output_3);
    assert!(output_3 == expected_3);
}
